### PR TITLE
Custom database path support

### PR DESCRIPTION
**Pull request by DemiMarie**
See individual commit messages for details.

### CODE REVIEW

**Comment by marmarek** [comment]
PipelineRetry

**Comment by marmarek** [comment]
```
DEBUG: error[E0433]: failed to resolve. Use of undeclared type or module `std`
DEBUG:   --> rpm-crypto/src/lib.rs:29:31
DEBUG:    |
DEBUG: 29 |     pub fn init(path: Option<&std::ffi::CStr>) -> InitToken {
DEBUG:    |                               ^^^ Use of undeclared type or module `std`
```

**Comment by marmarek** [comment]
Please also add
```
  - project: 'QubesOS/qubes-continuous-integration'
    file: '/r4.1/gitlab-vm.yml'
```
to `.gitlab-ci.yml`, to have the domU build tested too.

File: debian/control [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+Priority: optional
+Standards-Version: 5.0.0
+Homepage: https://www.qubes-os.org
+Vcs-Git: https://github.com/QubesOS/qubes-rpm-oxide.git
+Build-Depends:
+ rustc,
+ cargo,
+ librpm-dev,
+ debhelper,
+ debhelper-compat (= 13),
```
Comment (marmarek) [inline debian/control:13]: We do build packages for buster too, so this must be <= 12 (see CI).

File: debian/control [hunk 2]
```diff
@@ -0,0 +1,20 @@
+Source: qubes-rpm-oxide
+Maintainer: Demi Marie Obenour <demi@invisiblethingslab.com>
+Section: misc
+Priority: optional
+Standards-Version: 5.0.0
```
Comment (marmarek) [inline debian/control:5]: It seems the most recent policy version is 4.6.0.1 (https://www.debian.org/doc/debian-policy/index.html)

File: debian/control [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ rustc,
+ cargo,
+ librpm-dev,
+ debhelper,
+ debhelper-compat (= 13),
+Description: RPM canonicalization and verification tools
+
+Package: qubes-rpm-oxide
+Architecture: all
+Depends: ${misc:Depends}
```
Comment (marmarek) [inline debian/control:18]: Include `${shlibs:Depends}` too.

File: debian/rules [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/make -f
+
+export RUSTFLAGS=-Cdebuginfo=2 -Clink-arg=-z,relro,-z,now
+export RUSTC_BOOTSTRAP=1 
+
+build:
+	cargo build --all-features --release --offline
+install:
+	install -D -m 0755 -- target/release/rpmcanon debian/rpmcanon
```
Comment (marmarek) [inline debian/rules:9]:
```suggestion
	install -D -m 0755 -- target/release/rpmcanon debian/tmp/usr/bin/rpmcanon
```

File: debian/rules [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/make -f
+
+export RUSTFLAGS=-Cdebuginfo=2 -Clink-arg=-z,relro,-z,now
+export RUSTC_BOOTSTRAP=1 
+
+build:
```
Comment (marmarek) [inline debian/rules:6]:
```suggestion
override_dh_auto_build:
```

File: debian/rules [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+	cargo build --all-features --release --offline
+install:
```
Comment (marmarek) [inline debian/rules:8]:
```suggestion
override_dh_auto_install:
```

File: debian/rules [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+	install -D -m 0755 -- target/release/rpmcanon debian/rpmcanon
+clean:
```
Comment (marmarek) [inline debian/rules:10]:
```suggestion
override_dh_auto_clean:
```

File: rpm-writer/bin/rpmcanon.rs [hunk 8]
```diff
@@ -56,16 +56,62 @@ fn usage(success: bool) -> i32 {
...
+                },
+            }
+        }
+    }
+    impl std::io::Read for ProgressReportingReader {
+        fn read(&mut self, buffer: &mut [u8]) -> std::io::Result<usize> {
+            let bytes_read = self.inner.read(buffer)?;
+            if let Some(ref mut bytes_so_far) = self.do_report {
+                if bytes_read == 0 {
+                    println!("{} bytes total", *bytes_so_far);
```
Comment (marmarek) [inline rpm-writer/bin/rpmcanon.rs:88]: IMO better use consistent units - either both bytes, or both MiB. In either case, reporting progress only every 1MiB sounds fine.

File: rpm-crypto/src/lib.rs [hunk 9]
```diff
@@ -26,20 +26,43 @@ pub use init::{init, InitToken};
 mod init {
     #[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
     pub struct InitToken(());
-    pub fn init() -> InitToken {
+    pub fn init(path: Option<&std::ffi::CStr>) -> InitToken {
```
Comment (marmarek) [inline rpm-crypto/src/lib.rs:29]: This fails to build for R4.0.

File: rpm-crypto/src/lib.rs [hunk 10]
```diff
@@ -26,20 +26,43 @@ pub use init::{init, InitToken};
...
+        RPM_CRYPTO_INIT_ONCE.call_once(|| unsafe {
+            assert_eq!(rpmReadConfigFiles(ptr::null(), ptr::null()), 0);
+            if let Some(path) = path {
+                assert_eq!(
+                    rpmPushMacro(
+                        ptr::null_mut(),
+                        b"_dbpath\0".as_ptr() as _,
+                        ptr::null(),
+                        path.as_ptr(),
+                        -7
```
Comment (marmarek) [inline rpm-crypto/src/lib.rs:60]: Please at least add a comment that it's `RMIL_CMDLINE`. Or define constant.

**Review by marmarek — CHANGES_REQUESTED** [review]

**Comment by marmarek** [comment]
PipelineRetry

**Comment by DemiMarie** [comment]
PipelineRetry

**Comment by fepitre** [comment]
PipelineRefresh

**Comment by DemiMarie** [comment]
PipelineRetry

**Comment by DemiMarie** [comment]
@marmarek okay to merge?

**Review by marmarek — APPROVED** [review]

**Pull request merged**